File: Screenshot.lbaction/Contents/Scripts/imgcopy.m
#import <Foundation/Foundation.h>
#import <Cocoa/Cocoa.h>
#import <unistd.h>

// from http://www.alecjacobson.com/weblog/?p=3816

BOOL copy_to_clipboard(NSString *path)
{
  // http://stackoverflow.com/questions/2681630/how-to-read-png-image-to-nsimage
  NSImage * image;
  image =  [[NSImage alloc] initWithContentsOfFile:path];

  // http://stackoverflow.com/a/18124824/148668
  BOOL copied = false;
  if (image != nil) {
    NSPasteboard *pasteboard = [NSPasteboard generalPasteboard];
    [pasteboard clearContents];
    NSArray *copiedObjects = [NSArray arrayWithObject:image];
    copied = [pasteboard writeObjects:copiedObjects];
  }
  
  [image release];
  return copied;
}

int main(int argc, char * const argv[]) {
  NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];
  if(argc<1) {
    printf("Usage: Copy file to clipboard: ./imgcopy path/to/file\n\n");
    return EXIT_FAILURE;
  }
  NSString *path= [NSString stringWithUTF8String:argv[1]];
  BOOL success = copy_to_clipboard(path);
  [pool release];
  return (success?EXIT_SUCCESS:EXIT_FAILURE);
}

File: Screenshot.lbaction/Contents/Scripts/imgcopy.swift
// copy an image file to the clipboard
// inspired by http://www.alecjacobson.com/weblog/?p=3816
import Cocoa

let argCount = Process.arguments.count
if argCount > 1 {
  for path in Process.arguments[1..<argCount] {
    if let img = NSImage(contentsOfFile: path) {
      let clipboard = NSPasteboard.generalPasteboard()
      clipboard.clearContents()
      if !clipboard.writeObjects([img]) {
        print("Cannot copy file", path)
        exit(1)
      }
    } else {
      print("Cannot read file", path)
      exit(1)
    }
  }
}
exit(0)
